### PR TITLE
External Network: liqoctl network connect

### PR DESCRIPTION
**Pull request by fra98**
# Description

This PR adds the `liqoctl network connect` command. It also implements the REST interface for `GatewayServer`, `GatewayClient` and `PublicKey` resources

### CODE REVIEW

File: pkg/liqo-controller-manager/external-network/utils/getters.go [hunk 1]
```diff
@@ -124,3 +126,25 @@ func KindToResource(kind string) string {
 	// lowercased and pluralized
 	return strings.ToLower(kind) + "s"
 }
+
+// ResourceToKind returns the kind name for a given resource.
+func ResourceToKind(gvr schema.GroupVersionResource, config *rest.Config) (string, error) {
+	dClient, err := discovery.NewDiscoveryClientForConfig(config)
+	if err != nil {
+		return "", err
+	}
```
Comment (cheina97) [inline pkg/liqo-controller-manager/external-network/utils/getters.go:135]: Can you use the client inside the factory? It contains f.KubeClient.Discovery().ServerGroupsAndResources().

File: pkg/liqo-controller-manager/external-network/utils/getters.go [hunk 2]
```diff
@@ -124,3 +126,25 @@ func KindToResource(kind string) string {
...
+
+	res, err := dClient.ServerResourcesForGroupVersion(gvr.GroupVersion().String())
+	if err != nil {
+		return "", err
+	}
+
+	for i := range res.APIResources {
+		k := res.APIResources[i].Kind
+		if gvr.Resource == KindToResource(k) {
```
Comment (cheina97) [inline pkg/liqo-controller-manager/external-network/utils/getters.go:144]: https://github.com/kubernetes/apimachinery/blob/master/pkg/api/meta/restmapper.go#L126
Can you use this function? We already discussed about it but maybe we forgot to change it.

File: pkg/liqoctl/network/handler.go [hunk 3]
```diff
@@ -71,3 +103,110 @@ func (o *Options) RunInit(ctx context.Context) error {
...
+	if err != nil {
+		return err
+	}
+
+	// Create PublicKey of gateway client on cluster 1
+	if err := cluster1.EnsurePublicKey(ctx, cluster2.clusterIdentity, keyClient); err != nil {
+		return err
+	}
+
+	// TODO: Wait Connection on cluster 1 and cluster 2
```
Comment (cheina97) [inline pkg/liqoctl/network/handler.go:180]: Still required?

File: pkg/liqoctl/wait/wait.go [hunk 4]
```diff
@@ -222,3 +222,60 @@ func (w *Waiter) ForConfiguration(ctx context.Context, conf *networkingv1alpha1.
 	s.Success("Configuration applied successfully")
 	return nil
 }
+
+// ForGatewayServerStatusEndpoint waits until the service of a Gateway resource has been created
+// (i.e., until its endpoint status is not set).
+func (w *Waiter) ForGatewayServerStatusEndpoint(ctx context.Context, gwServer *networkingv1alpha1.GatewayServer) error {
+	s := w.Printer.StartSpinner("Waiting for gateway server endpoint status to be set")
```
Comment (cheina97) [inline pkg/liqoctl/wait/wait.go:229]: Maybe it's a too technical detail? What about printing it only with verbose?

File: pkg/liqoctl/rest/gatewayserver/create.go [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+	cmd.Flags().VarP(outputFormat, "output", "o",
+		"Output the resulting GatewayServer resource, instead of applying it. Supported formats: json, yaml")
+
+	cmd.Flags().StringVar(&o.ClusterID, "cluster-id", "", "The cluster ID of the remote cluster")
+	cmd.Flags().StringVar(&o.GatewayType, "type", DefaultGatewayType,
+		"Type of Gateway Server. Leave empty to use default Liqo implementation of WireGuard")
+	cmd.Flags().StringVar(&o.TemplateName, "template-name", DefaultTemplateName, "Name of the Gateway Server template")
+	cmd.Flags().StringVar(&o.TemplateNamespace, "template-namespace", DefaultTemplateNamespace, "Namespace of the Gateway Server template")
+	cmd.Flags().Var(o.ServiceType, "svc-type", fmt.Sprintf("Service type of Gateway Server. Default: %s", DefaultServiceType))
```
Comment (cheina97) [inline pkg/liqoctl/rest/gatewayserver/create.go:78]: What about using "service-type" like in the liqoctl install command?

Comment (cheina97) [reply pkg/liqoctl/rest/gatewayserver/create.go:78]: Same for liqoctl network connect

File: pkg/liqoctl/rest/gatewayserver/types.go [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gatewayserver
+
+import (
+	v1 "k8s.io/api/core/v1"
```
Comment (cheina97) [inline pkg/liqoctl/rest/gatewayserver/types.go:18]:
```suggestion
	corev1 "k8s.io/api/core/v1"
```

File: pkg/liqoctl/rest/gatewayserver/utils.go [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gatewayserver
+
+import (
+	v1 "k8s.io/api/core/v1"
```
Comment (cheina97) [inline pkg/liqoctl/rest/gatewayserver/utils.go:18]:
```suggestion
	corev1 "k8s.io/api/core/v1"
```

File: pkg/liqoctl/rest/publickey/create.go [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+		Run: func(cmd *cobra.Command, args []string) {
+			output.ExitOnErr(o.handleCreate(ctx))
+		},
+	}
+
+	cmd.Flags().VarP(outputFormat, "output", "o",
+		"Output the resulting PublicKey resource, instead of applying it. Supported formats: json, yaml")
+
+	cmd.Flags().StringVar(&o.ClusterID, "cluster-id", "", "The cluster ID of the remote cluster")
+	cmd.Flags().BytesBase64Var(&o.PublicKey, "public-key", nil, "The public key to be used for the Gateway")
```
Comment (cheina97) [inline pkg/liqoctl/rest/publickey/create.go:69]: A void array should be more safe?

**Review by cheina97 — APPROVED** [review]

File: pkg/liqoctl/rest/publickey/create.go (hunk 8 above)
Comment (fra98) [reply pkg/liqoctl/rest/publickey/create.go:69]: I would leave nil to differentiate between a nil slice and an empty slice. It should be safe as this flag is forwarded as-is to the PublicKey resource, so whoever uses it should handle it as they want.

**Pull request merged**